File: examples/kotlin/src/main/kotlin/adr/app/Assemble.kt
// ── app/assemble — the THREE total dispatchers (I3) ────────────────────────
//   foldApp            results → (state, effects)     the decision
//   projectApp         state   → AppView              what a human reads (6.9)
//   projectContextApp  state   → Context              what the reasoner reads (F4/G15)
//
// Three, not one, because they consume different things and 6.9 forbids fusing the
// fold with the projections.
//
// Every dispatcher is an EXHAUSTIVE match with NO else arm. Adding a block adds one
// branch to each, and the compiler names every one. Adding a VERB adds nothing here
// at all — that edit lands entirely inside the block (§11.1).
//
// THE BLOCKS ARE CONSTRUCTED HERE, per call. A block used to be a loose `object`, so
// there was nothing to build and nothing that could be built in a test; now each
// dispatcher stands its blocks up itself. The three signatures are UNCHANGED on
// purpose — `::foldApp` and `::projectContextApp` are passed to the Boundary against
// `typealias Fold<S>` and `ProjectContext<S>`, and an extra parameter here would break
// nine call sites. When these three dispatchers become a class in their own right, the
// locals below hoist to constructor state with no other rework.

package adr.app

import adr.blocks.analysis.AnalysisBlock
import adr.blocks.artifact.ArtifactBlock
import adr.blocks.console.ConsoleBlock
import adr.blocks.escalation.EscalationBlock
import adr.blocks.inbox.InboxBlock
import adr.blocks.triage.TriageBlock
import adr.contract.AnalysisResult
import adr.contract.ArtifactResult
import adr.contract.ConsoleResult
import adr.contract.Effect
import adr.contract.EscalationResult
import adr.contract.InboxResult
import adr.contract.ToolResult
import adr.contract.TriageResult
import adr.spine.pure.Context
import adr.spine.pure.MAX_CONTEXT_NOTICES
import adr.spine.pure.Notice
import adr.spine.pure.Signature
import adr.spine.pure.StagedInput
import adr.spine.pure.Timestamp
import adr.spine.pure.refusedArm
import adr.spine.pure.spineView
import adr.spine.pure.unhandledArm

fun foldApp(
    state: State,
    results: List<ToolResult>,
    now: Timestamp,
    sig: Signature,
): Pair<State, List<Effect>> {
    val triage = TriageBlock()
    val escalation = EscalationBlock()
    val console = ConsoleBlock()
    val artifact = ArtifactBlock()
    val analysis = AnalysisBlock()
    val inbox = InboxBlock()

    var s = state
    val effects = mutableListOf<Effect>()
    val notices = mutableListOf<Notice>()

    for (result in results) {
        when (result) {
            is TriageResult -> triage.arm(s.triage, result, now, sig).let {
                s = s.copy(triage = it.slice)
                effects += it.effects
                notices += it.notices
            }

            is EscalationResult -> escalation.arm(s.escalation, result, now, sig).let {
                s = s.copy(escalation = it.slice)
                effects += it.effects
                notices += it.notices
            }

            is ConsoleResult -> console.arm(s.console, result, now, sig).let {
                s = s.copy(console = it.slice)
                effects += it.effects
                notices += it.notices
            }

            is ArtifactResult -> artifact.arm(s.artifact, result, now, sig).let {
                s = s.copy(artifact = it.slice)
                effects += it.effects
                notices += it.notices
            }

            is AnalysisResult -> analysis.arm(s.analysis, result, now, sig).let {
                s = s.copy(analysis = it.slice)
                effects += it.effects
                notices += it.notices
            }

            is InboxResult -> inbox.arm(s.inbox, result, now, sig).let {
                s = s.copy(inbox = it.slice)
                effects += it.effects
                notices += it.notices
            }

            // The spine's own two arms. Identical everywhere (§7).
            is ToolResult.Unhandled -> unhandledArm(s.spine, result, now).let {
                effects += it.effects
                notices += it.notices
            }

            is ToolResult.Refused -> refusedArm(s.spine, result, now).let {
                effects += it.effects
                notices += it.notices
            }
        }
    }

    // Per-item notices land in the spine's slice. RunStatus is NEVER touched here (F9).
    return s.copy(spine = s.spine.withNotices(notices)) to effects.toList()
}

fun projectApp(state: State): AppView = AppView(
    root = spineView(state.spine),
    triage = TriageBlock().view(state.triage),
    escalation = EscalationBlock().view(state.escalation),
    console = ConsoleBlock().view(state.console),
    artifact = ArtifactBlock().view(state.artifact),
    analysis = AnalysisBlock().view(state.analysis),
    inbox = InboxBlock().view(state.inbox),
)

/**
 * The THIRD pure projection (G15). Recomputed from committed State every step, never
 * appended to, and bounded by declared constants — so |Context| is O(1) in timeline
 * length. The artifact contributes a COUNT, never its lines.
 */
fun projectContextApp(state: State, staged: List<StagedInput>): Context = Context(
    staged = staged,
    lines = TriageBlock().contextLines(state.triage) +
        EscalationBlock().contextLines(state.escalation) +
        ConsoleBlock().contextLines(state.console) +
        AnalysisBlock().contextLines(state.analysis) +
        InboxBlock().contextLines(state.inbox),
    notices = state.spine.notices
        .takeLast(MAX_CONTEXT_NOTICES)
        .map { "${it.tool.value}: ${it.reason}" },
    artifactLineCount = ArtifactBlock().lineCount(state.artifact),
)

File: examples/typescript/src/app/assemble.ts
// ── app/assemble — THE THREE TOTAL DISPATCHERS (I3) ────────────────────────
//   fold            results → (State, Effect[])      the only decider
//   project         State   → ViewModel              the Presenter (6.9)
//   projectContext  State   → Context                the reasoner's input (F4)
//
// Three, not one: they consume different things (results / slices / slices) and
// 6.9 forbids fusing the fold with the projections.
//
// EXHAUSTIVENESS, TWICE. `foldOne` closes over `outcome` with a `never` guard;
// `foldOk` closes over BLOCK OWNERSHIP with a `never` guard. TypeScript has no
// sealed sub-hierarchies, so the second one is a chain of type predicates
// rather than a `switch` — but the property is the same and it is enforced the
// same way: add a block's result type to `app/contract`'s union without adding
// its branch here and the build fails on `const _never: never = r`.
//
// Note what is NOT here: a per-VERB branch. Adding a verb touches four sites,
// all four inside its own block folder, and zero at the root (L5, §11.1).

import type { Signature } from "../spine/pure/actor";
import type { Context } from "../spine/pure/context";
import { MAX_CONTEXT_NOTICES, bounded } from "../spine/pure/context";
import type { Timestamp } from "../spine/pure/ids";
import type { Notice } from "../spine/pure/notice";
import { renderNotice } from "../spine/pure/notice";
import type { StagedInput } from "../spine/pure/staged";
import { spineArm, unclaimedArm, withNotices } from "../spine/pure/spine-slice";
import type { ToolResultBase } from "../spine/pure/tool-result";
import { spineView } from "../spine/pure/view";
import type { ArmOut, FoldOut } from "../spine/pure/verb";

import { analysis } from "../blocks/analysis/register";
import { artifact } from "../blocks/artifact/register";
import { consoleBlock } from "../blocks/console/register";
import { escalation } from "../blocks/escalation/register";
import { inbox } from "../blocks/inbox/register";
import { triage } from "../blocks/triage/register";

import type { AppView, Effect, OkResult, State, ToolResult } from "./contract";

export function fold(
  state: State,
  results: readonly ToolResult[],
  now: Timestamp,
  sig: Signature,
): FoldOut<State> {
  let current = state;
  const effects: Effect[] = [];
  for (const r of results) {
    const out = foldOne(current, r, now, sig);
    current = out.state;
    effects.push(...(out.effects as readonly Effect[]));
  }
  return { state: current, effects };
}

function foldOne(state: State, r: ToolResult, now: Timestamp, sig: Signature): FoldOut<State> {
  switch (r.outcome) {
    // The spine's own two arms, identical in every application (§7): a
    // diagnostic and a per-item notice; no transition, no domain effect.
    case "unhandled":
    case "refused": {
      const out = spineArm(state.spine, r, now);
      return { state: { ...state, spine: withNotices(out.slice, out.notices) }, effects: out.effects };
    }
    case "ok":
      return foldOk(state, r, now, sig);
    default: {
      const _never: never = r;
      return _never;
    }
  }
}

function foldOk(state: State, r: OkResult, now: Timestamp, sig: Signature): FoldOut<State> {
  if (triage.owns(r)) {
    return merge(triage.arm(state.triage, r, now, sig), (slice) => ({ ...state, triage: slice }));
  }
  if (escalation.owns(r)) {
    return merge(escalation.arm(state.escalation, r, now, sig), (slice) => ({ ...state, escalation: slice }));
  }
  if (consoleBlock.owns(r)) {
    return merge(consoleBlock.arm(state.console, r, now, sig), (slice) => ({ ...state, console: slice }));
  }
  if (artifact.owns(r)) {
    return merge(artifact.arm(state.artifact, r, now, sig), (slice) => ({ ...state, artifact: slice }));
  }
  if (analysis.owns(r)) {
    return merge(analysis.arm(state.analysis, r, now, sig), (slice) => ({ ...state, analysis: slice }));
  }
  if (inbox.owns(r)) {
    return merge(inbox.arm(state.inbox, r, now, sig), (slice) => ({ ...state, inbox: slice }));
  }
  // Every "ok" result belongs to exactly one block. Add a block to the union in
  // app/contract without adding its branch above, and this line fails to
  // compile — which is the edit list L5 promises, produced by the compiler.
  const _never: never = r;
  void _never;

  // COMPILE-TIME EXHAUSTIVENESS IS NOT RUNTIME TOTALITY. The chain above is
  // type predicates, and TypeScript TRUSTS a predicate it cannot verify — so a
  // block whose `owns` went stale narrows `r` to `never` here while a real
  // value flows through. Returning `_never` returned `undefined`, and the
  // caller died on `out.effects is not iterable`: a crash, from the one arm
  // 6.5 says must never crash.
  //
  // So the arm obeys 6.5 like every other: total AND observable. It folds an
  // explicit Unhandled marker and emits a diagnostic, exactly as an unknown
  // tool name does — never a silent drop, and never a thrown fold.
  // The fold does not mint a ToolResult to get here — C7 forbids that, and it is
  // right to: transport comes from a verb body or the boundary, never the fold.
  // It asks the spine for the arm instead.
  const out = unclaimedArm(state.spine, (r as ToolResultBase).tool, now);
  return { state: { ...state, spine: withNotices(out.slice, out.notices) }, effects: out.effects };
}

/** A block's arm returns its own slice plus per-item notices; the notices land
 *  in the SPINE's slice, so a block never has to know that notices exist as a
 *  cross-cutting concern — and never has a way to touch RunStatus (F9). */
function merge<S>(out: ArmOut<S>, put: (slice: S) => State): FoldOut<State> {
  const next = put(out.slice);
  return {
    state: { ...next, spine: withNotices(next.spine, out.notices as readonly Notice[]) },
    effects: out.effects,
  };
}

export function project(state: State): AppView {
  return {
    ...spineView(state.spine),
    triage: triage.view(state.triage),
    escalation: escalation.view(state.escalation),
    console: consoleBlock.view(state.console),
    artifact: artifact.view(state.artifact),
    analysis: analysis.view(state.analysis),
    inbox: inbox.view(state.inbox),
  };
}

export function projectContext(state: State, staged: readonly StagedInput[]): Context {
  return {
    staged,
    lines: [
      ...triage.contextLines(state.triage),
      ...escalation.contextLines(state.escalation),
      ...consoleBlock.contextLines(state.console),
      ...artifact.contextLines(state.artifact),
      ...analysis.contextLines(state.analysis),
      ...inbox.contextLines(state.inbox),
    ],
    notices: bounded(state.spine.notices.map(renderNotice), MAX_CONTEXT_NOTICES),
    // the artifact by COUNT, never its lines — this is the growth bound
    artifactLineCount: state.artifact.lines.length,
  };
}

/** The Dispatchers bundle the impure seam and the replay harness are given.
 *  Declared with METHOD syntax so the app's CLOSED unions satisfy the spine's
 *  open base signature: the spine hands the fold only results its own registry
 *  produced, plus the spine's own two cases. */
export const dispatchers = {
  fold(state: State, results: readonly ToolResult[], now: Timestamp, sig: Signature): FoldOut<State> {
    return fold(state, results, now, sig);
  },
  projectContext,
};
